onnx/test/shape_inference_test.py

```diff
@@ -3232,6 +3232,80 @@ def test_celu_function_output_shape(self):  # type: () -> None
         )
         self._assert_inferred(graph, [make_tensor_value_info('Y', TensorProto.FLOAT, (25, 48, 16, 16))])
 
+    def prepare_input_initializer_tensors(self, initializer_shape, input_shape):  # type: ignore
+        nodes = [make_node('Add', ['x', 'y'], 'z')]
+        if initializer_shape is None:
+            initializer = []  # type: ignore
+        else:
+            initializer = [make_tensor("x", TensorProto.FLOAT, initializer_shape, ()),  # type: ignore
+                make_tensor("y", TensorProto.FLOAT, initializer_shape, ())]
+        if input_shape is None:
+            inputs = []  # type: ignore
+        else:
+            inputs = [helper.make_tensor_value_info('x', TensorProto.FLOAT, input_shape),  # type: ignore
+                helper.make_tensor_value_info('y', TensorProto.FLOAT, input_shape)]
+
+        graph = helper.make_graph(nodes, "test", inputs=inputs, outputs=[], initializer=initializer, value_info=[])
+        return helper.make_model(graph)
+
+    def test_infer_with_initializer_without_input_above_ir4(self):  # type: () -> None
+        # This is for testing IR>=4: some tensors can only exist in initializer and not in input
+        # So shape_inference should make use of initializer shapes
+        initializer_shape = (8, 7)
+        original_model = self.prepare_input_initializer_tensors(initializer_shape, None)
+        inferred_model = onnx.shape_inference.infer_shapes(original_model)
+
+        # If shape inference fails, it will throw IndexError
+        z_tenor = inferred_model.graph.value_info.pop()
+        z_shape = (z_tenor.type.tensor_type.shape.dim[0].dim_value, z_tenor.type.tensor_type.shape.dim[1].dim_value)
+        assert z_shape == initializer_shape
+
+    def not_infer_empty_input_from_initializer(self, original_model):  # type: (ModelProto) -> None
+        # After shape inference, it won't update input from initializer
+        # So the inferred value info is empty
+
+        inferred_model = onnx.shape_inference.infer_shapes(original_model)
+        # Cannot infer from initializer; nothing in value_info; pop fail
+        self.assertRaises(IndexError, inferred_model.graph.value_info.pop)
+
+    def test_infer_with_initializer_without_input_below_ir4(self):  # type: () -> None
+        # This is for testing IR<4: tensors must exist both in initializer and input
+        # So shape_inference should not make use of initializer shapes
+        initializer_shape = (8, 7)
+        original_model = self.prepare_input_initializer_tensors(initializer_shape, None)
```

Comment:
In IR version 3, we expect every initializer to also be in input. Does this not cause an error? I think it should be better to use (None, None) as input_shape.

Reply:
Using ir_version=3 does not cause an error with current checker.
Using `None` here is to not pass input (empty list) to `make_tensor`.

Reply:
Hmmm. But it will create a model with x and y as initializer, but not as inputs, in IR version 3. I think it is working because you are calling only shape-inference here ... but the model will not pass this check in checker.cc https://github.com/onnx/onnx/blob/434cf85abde5b4eb3aa537bcbe892b39eed3e9c6/onnx/checker.cc#L679 , I think. So, isn't it an invalid model?

Reply:
You are right... It will be caught by `ValidationError: x in initializer but not in graph input`.
I will use (None, None) as input to produce a valid model. Just changed this test.
Thank you for the good catch!